sklearn/neural_network/tests/test_mlp.py

```diff
@@ -876,3 +876,20 @@ def test_mlp_loading_from_joblib_partial_fit(tmp_path):
     # finetuned model learned the new target
     predicted_value = load_estimator.predict(fine_tune_features)
     assert_allclose(predicted_value, fine_tune_target, rtol=1e-4)
+
+
+@pytest.mark.parametrize("Estimator", [MLPClassifier, MLPRegressor])
+def test_preserve_feature_names(recwarn, Estimator):
+    # Non-regression test for #24846
+
+    pd = pytest.importorskip("pandas")
+
+    X = pd.DataFrame(
+        data=[[i, i] for i in range(10)], columns=["colname_a", "colname_b"]
+    )
+    y = pd.DataFrame(data=[[1] for i in range(10)], columns=["colname_y"])
+
+    model = Estimator(early_stopping=True, validation_fraction=0.2)
+    model.fit(X, y["colname_y"])
+
+    assert len(recwarn) == 0
```

Comment:
I prefer converting the `UserWarning` into an error and letting it raise. This way the warning we are checking for is explicitly shown and the warning message is shown:

```python
    with warnings.catch_warnings():
        warnings.simplefilter("error", UserWarning)
        model.fit(X, y)
```

With this, the error message is shown:

```bash
        if X_feature_names is None and fitted_feature_names is not None:
>           warnings.warn(
                "X does not have valid feature names, but"
                f" {self.__class__.__name__} was fitted with feature names"
            )
E           UserWarning: X does not have valid feature names, but MLPRegressor was fitted with feature names

sklearn/base.py:409: UserWarning
```